File: python/identitykit/signing.py
"""Document signing, verification, and did:key codec for IdentityKit.

Ed25519 via the vetted `cryptography` library when present, falling back to the
pure-Python RFC 8032 reference (NOT constant-time) with a warning. The signing
key is a 32-byte seed that never leaves the caller. `did:key` uses standard
multibase base58btc with the ed25519 multicodec prefix, so identifiers are
interoperable with any did:key implementation. Zero third-party deps required.
"""

from __future__ import annotations

import base64
import os
import warnings
from typing import Tuple

from .canonical import canonicalize

try:
    from cryptography.hazmat.primitives.asymmetric.ed25519 import (
        Ed25519PrivateKey,
        Ed25519PublicKey,
    )

    _BACKEND = "cryptography"
except ImportError:  # pragma: no cover - exercised only without the dep
    from . import ed25519 as _ref

    _BACKEND = "pure-python"
    warnings.warn(
        "identitykit: 'cryptography' is not installed; falling back to a pure-Python "
        "Ed25519 reference that is NOT constant-time. Install 'cryptography' for "
        "production use.",
        RuntimeWarning,
        stacklevel=2,
    )


# -- raw Ed25519 -------------------------------------------------------------

def public_from_seed(seed: bytes) -> bytes:
    if _BACKEND == "cryptography":
        return Ed25519PrivateKey.from_private_bytes(seed).public_key().public_bytes_raw()
    return _ref.publickey(seed)


def sign_bytes(message: bytes, seed: bytes) -> bytes:
    if _BACKEND == "cryptography":
        return Ed25519PrivateKey.from_private_bytes(seed).sign(message)
    return _ref.sign(message, seed, _ref.publickey(seed))


def verify_bytes(signature: bytes, message: bytes, public_key: bytes) -> bool:
    if len(public_key) != 32 or len(signature) != 64:
        return False
    if _BACKEND == "cryptography":
        try:
            Ed25519PublicKey.from_public_bytes(public_key).verify(signature, message)
            return True
        except Exception:
            return False
    return _ref.verify(signature, message, public_key)


def generate_keypair() -> Tuple[bytes, bytes]:
    """Return (private_seed, public_key) as raw 32-byte values."""
    seed = os.urandom(32)
    return seed, public_from_seed(seed)


def b64(b: bytes) -> str:
    return base64.b64encode(b).decode("ascii")


def unb64(s: str) -> bytes:
    return base64.b64decode(s)


# -- base58btc (Bitcoin alphabet), for did:key multibase ---------------------

_B58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
_B58_INDEX = {c: i for i, c in enumerate(_B58)}


def _b58encode(data: bytes) -> str:
    n = int.from_bytes(data, "big")
    out = ""
    while n > 0:
        n, r = divmod(n, 58)
        out = _B58[r] + out
    pad = len(data) - len(data.lstrip(b"\x00"))
    return "1" * pad + out


def _b58decode(s: str) -> bytes:
    n = 0
    for ch in s:
        if ch not in _B58_INDEX:
            raise ValueError(f"invalid base58 character {ch!r}")
        n = n * 58 + _B58_INDEX[ch]
    body = n.to_bytes((n.bit_length() + 7) // 8, "big") if n else b""
    pad = len(s) - len(s.lstrip("1"))
    return b"\x00" * pad + body


# Multicodec prefix for an Ed25519 public key: 0xed 0x01 (varint).
_ED25519_MULTICODEC = b"\xed\x01"


def did_key_from_public(public_key: bytes) -> str:
    """Encode a raw Ed25519 public key as a did:key identifier."""
    if len(public_key) != 32:
        raise ValueError("ed25519 public key must be 32 bytes")
    return "did:key:z" + _b58encode(_ED25519_MULTICODEC + public_key)


def public_from_did_key(did: str) -> bytes:
    """Decode a did:key Ed25519 identifier back to its raw public key."""
    if not did.startswith("did:key:z"):
        raise ValueError("not an ed25519 did:key (expected 'did:key:z' multibase prefix)")
    # A valid ed25519 did:key body is ~48 chars; cap the input so a hostile,
    # oversized string can't force a pathological base58/bigint decode.
    if len(did) > 120:
        raise ValueError("did:key is implausibly long")
    decoded = _b58decode(did[len("did:key:z"):])
    if decoded[:2] != _ED25519_MULTICODEC:
        raise ValueError("did:key is not an Ed25519 key (wrong multicodec prefix)")
    pub = decoded[2:]
    if len(pub) != 32:
        raise ValueError("decoded public key is not 32 bytes")
    return pub


# -- detached proof over a canonical document --------------------------------

def make_proof(document: dict, seed: bytes, *, key_id: str = None) -> dict:
    """Sign the canonical form of `document` and return a detached proof object."""
    public_key = public_from_seed(seed)
    signature = sign_bytes(canonicalize(document), seed)
    proof = {
        "type": "Ed25519Signature2020",
        "public_key": b64(public_key),
        "value": b64(signature),
    }
    if key_id:
        proof["key_id"] = key_id
    return proof


def verify_proof(document: dict, proof: dict) -> bool:
    """True iff `proof` is a valid Ed25519 signature over `document`. Never raises."""
    if not isinstance(proof, dict) or proof.get("type") != "Ed25519Signature2020":
        return False
    try:
        public_key = unb64(proof["public_key"])
        value = unb64(proof["value"])
        return verify_bytes(value, canonicalize(document), public_key)
    except Exception:
        return False

File: typescript/src/signing.ts
/**
 * Document signing, verification, and the did:key codec, backed by Node's
 * built-in Ed25519 (no external crypto dependency). Keys are raw 32-byte values,
 * wire-compatible with the Python port: same canonical bytes, same signatures,
 * same base64. did:key uses standard multibase base58btc with the ed25519
 * multicodec prefix.
 */

import {
  createPrivateKey,
  createPublicKey,
  generateKeyPairSync,
  sign as nodeSign,
  verify as nodeVerify,
} from "node:crypto";
import { canonicalize } from "./canonical.ts";

const PKCS8_PREFIX = Buffer.from("302e020100300506032b657004220420", "hex");
const SPKI_PREFIX = Buffer.from("302a300506032b6570032100", "hex");

export interface Proof {
  type: "Ed25519Signature2020";
  public_key: string;
  value: string;
  key_id?: string;
}

export function generateKeypair(): { seed: Buffer; publicKey: Buffer } {
  const { privateKey, publicKey } = generateKeyPairSync("ed25519");
  const seed = privateKey.export({ format: "der", type: "pkcs8" }).subarray(-32);
  const pub = publicKey.export({ format: "der", type: "spki" }).subarray(-32);
  return { seed: Buffer.from(seed), publicKey: Buffer.from(pub) };
}

function privFromSeed(seed: Buffer) {
  return createPrivateKey({ key: Buffer.concat([PKCS8_PREFIX, seed]), format: "der", type: "pkcs8" });
}
function pubFromRaw(pub: Buffer) {
  return createPublicKey({ key: Buffer.concat([SPKI_PREFIX, pub]), format: "der", type: "spki" });
}

export function publicFromSeed(seed: Buffer): Buffer {
  const der = createPublicKey(privFromSeed(seed)).export({ format: "der", type: "spki" }).subarray(-32);
  return Buffer.from(der);
}

export function signBytes(message: Buffer, seed: Buffer): Buffer {
  return Buffer.from(nodeSign(null, message, privFromSeed(seed)));
}

export function verifyBytes(signature: Buffer, message: Buffer, publicKey: Buffer): boolean {
  if (publicKey.length !== 32 || signature.length !== 64) return false;
  try {
    return nodeVerify(null, message, pubFromRaw(publicKey), signature);
  } catch {
    return false;
  }
}

export const b64 = (b: Buffer): string => b.toString("base64");
export const unb64 = (s: string): Buffer => Buffer.from(s, "base64");

// -- base58btc (Bitcoin alphabet) --------------------------------------------

const B58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
const B58_INDEX: Record<string, number> = Object.fromEntries([...B58].map((c, i) => [c, i]));

function b58encode(data: Buffer): string {
  let n = 0n;
  for (const byte of data) n = n * 256n + BigInt(byte);
  let out = "";
  while (n > 0n) {
    const r = Number(n % 58n);
    n = n / 58n;
    out = B58[r] + out;
  }
  let pad = 0;
  for (const b of data) { if (b === 0) pad++; else break; }
  return "1".repeat(pad) + out;
}

function b58decode(s: string): Buffer {
  let n = 0n;
  for (const ch of s) {
    if (!(ch in B58_INDEX)) throw new Error(`invalid base58 character ${ch}`);
    n = n * 58n + BigInt(B58_INDEX[ch]);
  }
  const bytes: number[] = [];
  while (n > 0n) { bytes.unshift(Number(n % 256n)); n = n / 256n; }
  let pad = 0;
  for (const ch of s) { if (ch === "1") pad++; else break; }
  return Buffer.from([...new Array(pad).fill(0), ...bytes]);
}

const ED25519_MULTICODEC = Buffer.from([0xed, 0x01]);

export function didKeyFromPublic(publicKey: Buffer): string {
  if (publicKey.length !== 32) throw new Error("ed25519 public key must be 32 bytes");
  return "did:key:z" + b58encode(Buffer.concat([ED25519_MULTICODEC, publicKey]));
}

export function publicFromDidKey(did: string): Buffer {
  if (!did.startsWith("did:key:z")) throw new Error("not an ed25519 did:key");
  // Cap input length so a hostile oversized string can't force a pathological
  // base58/bigint decode (a valid ed25519 did:key body is ~48 chars).
  if (did.length > 120) throw new Error("did:key is implausibly long");
  const decoded = b58decode(did.slice("did:key:z".length));
  if (decoded[0] !== 0xed || decoded[1] !== 0x01) throw new Error("did:key is not Ed25519");
  const pub = decoded.subarray(2);
  if (pub.length !== 32) throw new Error("decoded public key is not 32 bytes");
  return Buffer.from(pub);
}

// -- detached proof ----------------------------------------------------------

export function makeProof(document: unknown, seed: Buffer, keyId?: string): Proof {
  const proof: Proof = {
    type: "Ed25519Signature2020",
    public_key: b64(publicFromSeed(seed)),
    value: b64(signBytes(canonicalize(document), seed)),
  };
  if (keyId) proof.key_id = keyId;
  return proof;
}

export function verifyProof(document: unknown, proof: unknown): boolean {
  if (!proof || typeof proof !== "object") return false;
  const p = proof as Proof;
  if (p.type !== "Ed25519Signature2020") return false;
  try {
    return verifyBytes(unb64(p.value), canonicalize(document), unb64(p.public_key));
  } catch {
    return false;
  }
}
